File: runtime/CSharp/Antlr4.Runtime/Atn/SetTransition.cs
// Copyright (c) Terence Parr, Sam Harwell. All Rights Reserved.
// Licensed under the BSD License. See LICENSE.txt in the project root for license information.

using Antlr4.Runtime.Misc;
using Antlr4.Runtime.Sharpen;

namespace Antlr4.Runtime.Atn
{
    /// <summary>A transition containing a set of values.</summary>
    public class SetTransition : Transition
    {
        [NotNull]
        public readonly IntervalSet set;

        public SetTransition(ATNState target, IntervalSet set)
            : base(target)
        {
            // TODO (sam): should we really allow null here?
            if (set == null)
            {
                set = IntervalSet.Of(TokenConstants.InvalidType);
            }
            this.set = set;
        }

        public override Antlr4.Runtime.Atn.TransitionType TransitionType
        {
            get
            {
                return Antlr4.Runtime.Atn.TransitionType.Set;
            }
        }

        public override IntervalSet Label
        {
            get
            {
                return set;
            }
        }

        public override bool Matches(int symbol, int minVocabSymbol, int maxVocabSymbol)
        {
            return set.Contains(symbol);
        }

        [return: NotNull]
        public override string ToString()
        {
            return set.ToString();
        }
    }
}

File: reference/antlr4/runtime/Java/src/org/antlr/v4/runtime/atn/SetTransition.java
/*
 * Copyright (c) 2012 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD-3-Clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

package org.antlr.v4.runtime.atn;

import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.misc.IntervalSet;
import org.antlr.v4.runtime.misc.NotNull;
import org.antlr.v4.runtime.misc.Nullable;

/** A transition containing a set of values. */
public class SetTransition extends Transition {
	@NotNull
	public final IntervalSet set;

	// TODO (sam): should we really allow null here?
	public SetTransition(@NotNull ATNState target, @Nullable IntervalSet set) {
		super(target);
		if ( set == null ) set = IntervalSet.of(Token.INVALID_TYPE);
		this.set = set;
	}

	@Override
	public TransitionType getSerializationType() {
		return TransitionType.SET;
	}

	@Override
	@NotNull
	public IntervalSet label() { return set; }

	@Override
	public boolean matches(int symbol, int minVocabSymbol, int maxVocabSymbol) {
		return set.contains(symbol);
	}

	@Override
	@NotNull
	public String toString() {
		return set.toString();
	}
}
